pkg/controller/federatedcluster/controller.go

```diff
@@ -160,82 +150,146 @@ func (cc *ClusterController) Run(stopChan <-chan struct{}) {
 		if err := cc.updateClusterStatus(); err != nil {
 			glog.Errorf("Error monitoring cluster status: %v", err)
 		}
-	}, cc.clusterMonitorPeriod, stopChan)
+	}, cc.clusterHealthCheckConfig.PeriodSeconds, stopChan)
 }
 
-// updateClusterStatus checks cluster status and get the metrics from cluster's restapi
+// updateClusterStatus checks cluster health and updates status of all FederatedClusters
 func (cc *ClusterController) updateClusterStatus() error {
 	clusters := &fedv1a1.FederatedClusterList{}
 	err := cc.client.List(context.TODO(), clusters, cc.fedNamespace)
 	if err != nil {
 		return err
 	}
 
-	for _, cluster := range clusters.Items {
+	var wg sync.WaitGroup
+	for _, obj := range clusters.Items {
 		cc.mu.RLock()
-		// skip updating status of the cluster which is not yet added to knownClusterSet.
-		if !cc.knownClusterSet.Has(cluster.Name) {
-			cc.mu.RUnlock()
+		cluster := obj.DeepCopy()
+		clusterStore := cc.clusterStoreMap[cluster.Name]
+		cc.mu.RUnlock()
+		if clusterStore == nil {
+			glog.Warningf("Failed to retrieve stored data for cluster %s", cluster.Name)
 			continue
 		}
-		clusterClient, clientFound := cc.clusterKubeClientMap[cluster.Name]
-		clusterStatusOld, statusFound := cc.clusterStatusMap[cluster.Name]
-		cc.mu.RUnlock()
 
-		if !clientFound {
-			glog.Warningf("Failed to get client for cluster %s", cluster.Name)
-			continue
+		wg.Add(1)
+		go cc.updateIndividualClusterStatus(cluster, clusterStore, &wg)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (cc *ClusterController) updateIndividualClusterStatus(cluster *fedv1a1.FederatedCluster,
+	store *ClusterStore, wg *sync.WaitGroup) {
+	clusterClient := store.clusterKubeClient
+
+	currentClusterStatus := thresholdAdjustedClusterStatus(clusterClient.GetClusterHealthStatus(), store,
+		cc.clusterHealthCheckConfig)
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CrossClusterServiceDiscovery) {
+		currentClusterStatus = updateClusterZoneAndRegion(currentClusterStatus, store, clusterClient, cluster.Name)
+	}
+
+	store.clusterStatus = currentClusterStatus
+	cluster.Status = *currentClusterStatus
+	if err := cc.client.UpdateStatus(context.TODO(), cluster); err != nil {
+		glog.Warningf("Failed to update the status of cluster: %v, error is : %v", cluster.Name, err)
+	}
+	wg.Done()
+}
+
+func thresholdAdjustedClusterStatus(clusterStatus *fedv1a1.FederatedClusterStatus, store *ClusterStore,
+	clusterHealthCheckConfig util.ClusterHealthCheckConfig) *fedv1a1.FederatedClusterStatus {
+
+	if store.clusterStatus == nil {
+		return clusterStatus
+	}
+
+	threshold := clusterHealthCheckConfig.FailureThreshold
+	if isClusterReady(clusterStatus) {
+		threshold = clusterHealthCheckConfig.SuccessThreshold
+	}
+
+	if store.resultRun < threshold {
```

Comment:
i guess you should update resultRun before checking threshold.

Reply:
Taken care now.

Reply:
I think it was a mistake and messes up the logic. Have updated PTAL again.